### PR TITLE
Added validation for all types of cfbs.json files and all fields

### PR DESCRIPTION
**Pull request by olehermanse**

### CODE REVIEW

**Review by larsewi — COMMENTED** [review]
Great work 🚀 Just a quick note: I've noticed that you occasionally leave the code in a broken state in one commit and later fix it in another commit, presumably when you discovered it. Ideally, these two commits should be squashed. This makes it easier to revert changes in the future. It also makes it easier to use `git bisect` to identify the commit that introduced a bug. And last but not least, it hides the fact that you are human, capable of making mistakes. This will increase your chances of surviving when AI takes over. But it's not a big deal, just something to keep in mind.

File: cfbs/validate.py [hunk 1]
```diff
@@ -199,7 +199,7 @@ def validate_url_field(name, module, field):
         required_fields.append("version")
         required_fields.append("commit")
 
-    for required_field in required_field:
+    for required_field in required_fields:
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply cfbs/validate.py:412]: Tried, does not apply without conflicts. IMO in these cases, as long as the commits have good commit messages, it should be fine.

File: cfbs/cfbs_config.py [hunk 2]
```diff
@@ -485,3 +485,30 @@ def _input_list(input_data):
...
+            modules.extend(self.index.keys())
+
+        return modules
+
+    def can_reach_dependency(self, name, search_in=("build", "provides", "index")):
+        return name in self._get_all_module_names()
+
+    def find_module(self, name, search_in=("build", "provides", "index")):
+        if "build" in search_in and "build" in self:
+            for module in (x for x in self["build"]):
```
Comment (larsewi) [inline cfbs/cfbs_config.py:506]: Is not this the same? 🤔 
```suggestion
            for module in self["build"]:
```

Comment (olehermanse) [reply cfbs/cfbs_config.py:506]: 👍 Yea, for a list it is, for an array it would not be.

File: tests/test_validate_index_alias.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    with open(filename, "w") as f:
+        f.write(pretty(INDEX_ALIAS_VALID))
+    config = CFBSConfig(filename)
+    r = validate_config(config)
+    del config
+    filename.unlink()
+    assert r == 0
+
+
+def test_valid_index_with_alias():
```
Comment (craigcomstock) [inline tests/test_validate_index_alias.py:53]: maybe `test_invalid_index_with_alias()`?

**Review by craigcomstock — COMMENTED** [review]
after https://github.com/cfengine/cfbs/pull/167/commits/5f576e383b00ef7b459fe7f93f08f066f733e416 looks good :+1:

File: cfbs/validate.py [hunk 4]
```diff
@@ -2,188 +2,483 @@
...
                 )
-
-    def validate_url_field(name, modules, field):
-        url = modules[name].get(field)
+            step_array = step.split(" ")
+            operation, args = step_array[0], step_array[1:]
+            if not operation in AVAILABLE_BUILD_STEPS:
+                x = ", ".join(AVAILABLE_BUILD_STEPS)
+                raise CFBSValidationError(
+                    name, 'Unknown operation in "steps", must be one of: (%s)' % x
```
Comment (craigcomstock) [inline cfbs/validate.py:274]:
```suggestion
                    name, 'Unknown operation "%s" in "steps", must be one of: (%s)' % (operation, x)
```

File: cfbs/validate.py [hunk 5]
```diff
@@ -2,188 +2,483 @@
...
+            if input_element["type"] not in ("string", "list"):
+                raise CFBSValidationError(
+                    name,
+                    'The input "type" must be "string" or "list", not "%s"'
+                    % input_element["type"],
+                )
+            if not re.fullmatch(r"[a-z_]+", input_element["variable"]):
+                raise CFBSValidationError(
+                    name,
+                    '"%s" is not an acceptable variable name'
```
Comment (craigcomstock) [inline cfbs/validate.py:344]:
```suggestion
                    '"%s" is not an acceptable variable name, must match regex "[a-z_]+"'
```

File: cfbs/validate.py [hunk 6]
```diff
@@ -2,188 +2,483 @@
...
+                    % input_element["variable"],
+                )
+            if not re.fullmatch(r"[a-z_]+", input_element["namespace"]):
+                raise CFBSValidationError(
+                    name,
+                    '"%s" is not an acceptable namespace' % input_element["namespace"],
```
Comment (craigcomstock) [inline cfbs/validate.py:350]:
```suggestion
                    '"%s" is not an acceptable namespace, must match regex "[a-z_]+"' % input_element["namespace"],
```

File: cfbs/validate.py [hunk 7]
```diff
@@ -2,188 +2,483 @@
...
+                )
+            if not re.fullmatch(r"[a-z_]+", input_element["bundle"]):
+                raise CFBSValidationError(
+                    name,
+                    '"%s" is not an acceptable bundle name' % input_element["bundle"],
```
Comment (craigcomstock) [inline cfbs/validate.py:355]:
```suggestion
                    '"%s" is not an acceptable bundle name, must match regex "[a-z_]+"' % input_element["bundle"],
```

**Review by craigcomstock — COMMENTED** [review]
A few warning message suggestions. Looks good.

File: cfbs/validate.py [hunk 8]
```diff
@@ -2,188 +2,487 @@
...
+                            )
+                        if type(part[field]) is not str or part[field].strip() == "":
+                            raise CFBSValidationError(
+                                name,
+                                'The "%s" field in module input "subtype" objects must be a non-empty / non-whitespace string'
+                                % field,
+                            )
+                    if len(subtype) > 1:
+                        # The "key" field is used to create the JSON objects for each
+                        # input in  a list of "things" which are not just strings,
```
Comment (larsewi) [inline cfbs/validate.py:403]:
```suggestion
                        # input in a list of "things" which are not just strings,
```

**Review by larsewi — APPROVED** [review]

**Pull request merged**